Sources/SwiftCrossUI/Views/NavigationPath.swift

```diff
@@ -1,33 +1,150 @@
+import Foundation
+
 /// A type-erased list of data representing the content of a navigation stack.
+///
+/// If you are persisting a path using the ``Codable`` implementation, you must
+/// not change type definitions in a non-backwards compatible way. Otherwise,
+/// the path may fail to decode.
 public struct NavigationPath {
-    /// The current navigation path.
-    private(set) var path: [any Hashable] = []
+    /// A storage class used so that we have control over exactly which changes are published (to
+    /// avoid infinite loops).
+    private class Storage {
+        /// An entry that will be decoded next time it is used by a ``NavigationStack`` (we need to
+        /// wait until we know what concrete entry types are available).
+        struct EncodedEntry: Codable {
+            var type: String
+            var value: Data
+        }
 
-    /// A Boolean that indicates whether this path is empty.
-    public var isEmpty: Bool {
-        path.isEmpty
+        /// The current path. If both this and `encodedEntries` are non-empty, the elements in path
+        /// were added before the navigation path was even used to render a view. By design they
+        /// come after the encodedEntries (because they can only be the result of appending and
+        /// maybe popping).
+        var path: [any Codable] = []
+        /// Entries that will be encoded when this navigation path is first used by a
+        /// ``NavigationStack``. It is not possible to decode the entries without first knowing
+        /// what types the path can possibly contain (which only the ``NavigationStack`` will know).
+        var encodedEntries: [EncodedEntry] = []
     }
 
-    /// The number of elements in this path.
-    public var count: Int {
-        path.count
+    /// The path and any elements waiting to be decoded are stored in a class so that changes are
+    /// triggered from within NavigationStack when decoding the elements (which causes an infinite
+    /// loop of updates).
+    private var storage = Storage()
+
+    /// Indicates whether this path is empty.
+    var isEmpty: Bool {
+        storage.encodedEntries.isEmpty && storage.path.isEmpty
+    }
+
+    /// The number of elements in the path.
+    var count: Int {
+        storage.encodedEntries.count + storage.path.count
     }
 
     /// Creates an empty navigation path.
     public init() {}
 
-    /// Appends a new value to the end of this path.
-    public mutating func append<H: Hashable>(_ component: H) {
-        path.append(component)
+    /// Appends a new value to the end of the path.
+    public mutating func append<C: Codable>(_ component: C) {
```

Comment:
I dont mind either way but Apples SwiftUI does it this way and the WWDC video on any says to use generics instead of any where possible: 
![image](https://github.com/stackotter/swift-cross-ui/assets/17001151/69bb2639-02e9-487e-9fd7-d8221ab22575)
https://developer.apple.com/videos/play/wwdc2022/110354/?time=2130 (35:30)